client/concordclient/src/concord_client.cpp

```diff
@@ -31,6 +35,75 @@ ConcordClient::ConcordClient(const ConcordClientConfig& config)
   auto trc_config = std::make_unique<ThinReplicaClientConfig>(
       config_.subscribe_config.id, trc_queue_, config_.topology.f_val, std::move(trs_connections));
   trc_ = std::make_unique<ThinReplicaClient>(std::move(trc_config), metrics_);
+  ConcordClientPoolConfig client_pool_config = createClientPoolStruct(config);
+  client_pool_ = std::make_unique<concord::concord_client_pool::ConcordClientPool>(client_pool_config, metrics_);
+  if (client_pool_->HealthStatus() == concord::concord_client_pool::PoolStatus::NotServing) {
+    LOG_ERROR(KEY_EX_LOG, "Client pool health status is Not Serving");
+    throw concord::concord_client_pool::InternalError();
+  }
+}
+
+ConcordClientPoolConfig ConcordClient::createClientPoolStruct(const ConcordClientConfig& config) {
+  ConcordClientPoolConfig client_pool_config;
+  int id = 0;
+  for (const auto& replica : config_.topology.replicas) {
+    Replica client_pool_replica;
+    auto replica_id = replica.id.val;
+    client_pool_replica.principal_id = replica_id;
+    client_pool_replica.replica_host = replica.host;
+    client_pool_replica.replica_port = replica.bft_port;
+    client_pool_config.node[id] = client_pool_replica;
+    id++;
+  }
+
+  ParticipantNode client_pool_pn;
+  id = 0;
+  for (const auto& bft_client : config_.bft_clients) {
+    ExternalClient client_pool_ec;
+    auto external_client_id = bft_client.id.val;
+    client_pool_ec.principal_id = external_client_id;
+    client_pool_ec.client_port = bft_client.port;
+    client_pool_pn.externalClients[id] = client_pool_ec;
```

Comment:
There is not documentation in the header for this field but do you know if the key `id` is relevant? Is it the bft client id? If it isn't then I'm wondering why `externalClients` isn't simply a `std::vector` 🤔

Reply:
As far as I understood, the id in the `externalClients` map is for mapping between the serialized number to principal id because we don't know the principal id, hence we are creating an id of our own.

Reply:
So, if we are creating our own then why don't we use `external_client_id`/`bft_client.id.val` instead?

Reply:
Because we got an order inside the replicas, the IDs of the clients must be greater than replica ids + RO replicas + internal clients.